sdks/java/core/src/main/java/org/apache/beam/sdk/util/RowFilter.java

```diff
@@ -0,0 +1,478 @@
+ *
+ * // this filter will drop these fields
+ * RowFilter droppingFilter = new RowFilter(beamSchema).dropping(fields);
+ *
+ * // this filter will unnest fields "abc" and "xyz" to the top level
+ * List<String> fields = Arrays.asList("foo.abc", "bar.baz.xyz");
+ * RowFilter unnestingFilter = new RowFilter(beamSchema).unnesting(fields);
+ *
+ * // produces a filtered row
+ * Row outputRow = keepingFilter.filter(row);
+ * }</pre>
+ *
+ * Check the documentation for {@link #keeping(List)}, {@link #dropping(List)}, and {@link
+ * #unnesting(List)} for further details on what an output Row can look like.
+ */
+public class RowFilter implements Serializable {
+  private final Schema rowSchema;
+  private @Nullable Schema transformedSchema;
+
+  public RowFilter(Schema rowSchema) {
+    this.rowSchema = rowSchema;
+  }
+
+  /**
+   * Configures this {@link RowFilter} to filter {@link Row}s by keeping only the specified fields.
+   * Nested fields can be specified using dot-notation.
+   *
+   * <p>For example, if we want to keep the list of fields {@code ["foo", "baz"]}, for the input
+   * {@link Row}:
+   *
+   * <pre>{@code
+   * foo: 123
+   * bar: 456
+   * baz:
+   *   nested_1: abc
+   *   nested_2: xyz
+   * }</pre>
+   *
+   * we will get the following output {@link Row}:
+   *
+   * <pre>{@code
+   * foo: 123
+   * baz
+   *   nested_1: abc
+   *   nested_2: xyz
+   * }</pre>
+   */
+  public RowFilter keeping(List<String> fields) {
+    checkUnconfigured();
+    verifyNoNestedFields(fields, "keep");
+    validateSchemaContainsFields(rowSchema, fields, "\"keep\"");
+    transformedSchema = keepFields(rowSchema, fields);
+    return this;
+  }
+
+  /**
+   * Configures this {@link RowFilter} to filter {@link Row} by removing the specified fields.
+   * Nested fields can be specified using dot-notation.
+   *
+   * <p>For example, if we want to drop the list of fields {@code ["foo", "baz"]}, for this input
+   * {@link Row}:
+   *
+   * <pre>{@code
+   * foo: 123
+   * bar: 456
+   * baz:
+   *   nested_1: abc
+   *   nested_2: xyz
+   * }</pre>
+   *
+   * we will get the following output {@link Row}:
+   *
+   * <pre>{@code
+   * bar: 456
+   * }</pre>
+   */
+  public RowFilter dropping(List<String> fields) {
+    checkUnconfigured();
+    verifyNoNestedFields(fields, "drop");
+    validateSchemaContainsFields(rowSchema, fields, "\"drop\"");
+    transformedSchema = dropFields(rowSchema, fields);
+    return this;
+  }
+
+  /**
+   * Configures this {@link RowFilter} to unnest the specified fields to the top-level and keeping
+   * their leaf names. The unnested fields are kept and everything else is dropped. This will fail
+   * if two fields have identical leaf names. Nested fields can be specified using dot-notation.
+   *
+   * <p>For example, if we want to unnest the list of fields {@code ["abc", "foo.bar",
+   * "foo.xyz.baz"]}, for this input {@link Row}:
+   *
+   * <pre>{@code
+   * abc: 123
+   * foo:
```

Comment:
So I think what will be useful here is to indicate that one only wants to write foo, but not nested as 

```
foo:
  bar: my_str
  xyz:
    baz: 456
    que: 789
```

but to a sink expecting Foo type and the records are

```
bar: my_str
xyz:
  baz: 456
  que: 789
```

(e.g. I think it'd be quite common to have elements {path: ..., record: ...} and want to write out record. 

To use `unnest` as written one would have to enumerate all the fields of `foo` (or `record`). This is why I was prosing naming this `only` and making there be exactly one.

Reply:
Ahh I see, that does look like a better approach. 

Does it make sense to make it a list option? It would server the purpose of `only` but also allow unnesting multiple rows. Just wondering because it would be inconvenient in the future if we ever wanted to turn it from a `string` to a `list<string>`

Reply:
P.S. should we also limit `unnest` to top-level fields? 

ie. in your example one can do `unnest: foo` but not `unnest: foo.xyz`

Reply:
Yes, let's limit to top-level fields as well (for now at least). 

For naming, I think it'd be useful to get lots of eyes/opinions. Created https://docs.google.com/document/d/1IIn4cjF9eYASnjSmVmmAt6ymFnpBxHgBKVPgpnQ12G4/edit

Reply:
If there's no support for a list `unnest` option by Monday, I'm down to settle on a string `only` option. After thinking a lil about it, it doesn't make too much sense to unnest multiple records, effectively merging their contents into one bigger record.

Reply:
Just pushed a commit to switch to the `only` option. PTAL